### PR TITLE
[Mono.Compiler] use generalized execute entry via runtime_invoke wrapper

### PR DESCRIPTION
**Pull request by lewurm**
we need the method signature for that somewhow, so I attach MethodInfo
to NativeCodeHandle.

### CODE REVIEW

File: mcs/class/Mono.Compiler/Test/ICompilerInterfaceTest.cs [hunk 1]
```diff
@@ -64,12 +101,12 @@ public class ICompilerTests
 
 			var move1 = it.MoveNext ();
 			Assert.IsTrue (move1);
-			Assert.AreEqual (Opcode.Ldarg1, it.Opcode, "instr 1");
+			Assert.AreEqual (Opcode.Ldarg0, it.Opcode, "instr 1");
```
Comment (lambdageek) [inline mcs/class/Mono.Compiler/Test/ICompilerInterfaceTest.cs]: ? Is that right?  this is what I saw in `monodis`:

```
   .method public hidebysig 
           instance default int32 AddMethod (int32 a, int32 b)  cil managed 
    {
        // Method begins at RVA 0x2068
        // Code size 4 (0x4)
        .maxstack 8
        IL_0000:  ldarg.1 
        IL_0001:  ldarg.2 
        IL_0002:  add 
        IL_0003:  ret 
    } // end of method ICompilerTests::AddMethod
```

File: mono/mini/mini-runtime.c [hunk 2]
```diff
@@ -4548,41 +4548,97 @@ typedef struct _InstalledRuntimeCode InstalledRuntimeCode;
...
+
+	if (!MethodInfo_klass) {
+		MethodInfo_klass = mono_class_from_name_checked (mono_defaults.compiler, "Mono.Compiler", "MethodInfo", error);
+		return_val_if_nok (error, NULL);
+
+		g_assert (MethodInfo_klass);
+	}
+
+	if (!MethodInfo_get_RuntimeMethodHandle_method) {
+		MethodInfo_get_RuntimeMethodHandle_method = mono_class_get_method_from_name_checked (MethodInfo_klass, "get_RuntimeMethodHandle", 0, 0, error);
```
Comment (lambdageek) [inline mono/mini/mini-runtime.c]: we could add a real backing field and a struct on the C side.

File: mcs/class/Mono.Compiler/Test/ICompilerInterfaceTest.cs (hunk 1 above)
Comment (lambdageek) [reply mcs/class/Mono.Compiler/Test/ICompilerInterfaceTest.cs]: oh `AddMethod` is static now. my bad.

Comment (lewurm) [reply mcs/class/Mono.Compiler/Test/ICompilerInterfaceTest.cs]: I changed it to `static`

File: mono/mini/mini-runtime.c (hunk 2 above)
Comment (lewurm) [reply mono/mini/mini-runtime.c]: so I tried that first and somehow I failed miserably.  I didn't investigate any further as it worked with this approach just fine, but yes, it's a bit expensive.

**Comment by luhenry** [comment]
Adding the `MethodInfo` to the `NativeCodeHandle` looks weird. That is not something I would like to keep in the long run, but as long as it allows you to go forward for this hack week we should keep it.

For future reference, what's the reason you cannot use the `MethodInfo` passed to `CompileMethod`?

**Comment by lewurm** [comment]
@luhenry I agree, it's ugly. We need the corresponding `MethodInfo` on the native side somehow. What do you think about extending `IRuntimeInformation::InstallCompilationResult` by an `MethodInfo` object?

**Pull request merged**